MaxText/layers/linears.py

```diff
@@ -429,37 +436,64 @@ def generate_masks(self, top_k_indices):
       expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts, dtype=jnp.int32)
       reshaped_expert_mask = jnp.reshape(expert_mask, (batch_size, seq_len * self.num_experts_per_tok, self.num_experts))
       expert_token_count = jnp.cumsum(reshaped_expert_mask, axis=1)
-      expert_token_count = jnp.reshape(expert_token_count, ((batch_size, seq_len, self.num_experts_per_tok, self.num_experts)))
-      trunc_expert_mask = expert_mask * jnp.less_equal(expert_token_count, expert_capacity)
+      reshaped_expert_token_count = jnp.reshape(expert_token_count, ((batch_size, seq_len, self.num_experts_per_tok, self.num_experts)))
+      trunc_expert_mask = expert_mask * jnp.less_equal(reshaped_expert_token_count, expert_capacity_per_batch)
       updated_expert_mask = jnp.sum(trunc_expert_mask, axis=3)
-      return updated_expert_mask
+
+      # update weights for all experts
+      softmax_probs *= updated_expert_mask
+      weights = self.update_weights(softmax_probs, top_k_indices)
```

Comment:
Is this necessary here? Isn't softmax_probs already 0 in every entry not correpsonding to a top_k_index because we just multiplied it with updated_expert_mask?

Reply:
Yes, this is needed. The update_weights will do 2 things:

1) update weights due to the dropping;
2) reshape the weights from (batch, seq, topk_exp) to (batch, seq, num_exp).